yesod-core/Yesod/Core/Types.hs

```diff
@@ -308,6 +308,18 @@ newtype RepXml = RepXml Content
 
 type ContentType = ByteString -- FIXME Text?
 
+-- | Wrapper around types so that Handlers can be typed them, encoded as JSON.
+-- Example usage in a type signature:
+-- 
+-- > postSignupR :: Handler (JSONResponse CreateUserResponse)
+--
+-- And in the implementation:
+--
+-- > return $ JSONResponse $ CreateUserResponse userId
+--
+-- @since 1.6.3 
+data JSONResponse a = ToJSON a => JSONResponse a
```

Comment:
I think it would be better without the datatype context `ToJSON a =>` here. I don't remember all of the arguments, but it's typically considered a bad extension.

Reply:
Hm, this gives some reasons https://wiki.haskell.org/Data_declaration_with_constraint

Haven’t tried this, but the wiki recommends replacing with a GADT

Reply:
I don't think a GADT is necessary here. I'd replace with just:

```haskell
newtype JSONResponse a = JSONResponse a
```

Is anything lost in such a change?

Reply:
Does the GADT approach solve that?